libmamba/src/download/downloader_impl.hpp

```diff
@@ -105,9 +103,77 @@ namespace mamba
         std::unique_ptr<Impl> p_impl = nullptr;
     };
 
+    /*
+     * MirrorAttempt
+     *
+     * Manages the sequence of requests required to download an artifact
+     * on a mirror.
+     *
+     */
+    class MirrorAttempt
+    {
+    public:
+
+        using completion_function = DownloadAttempt::completion_function;
+        using on_success_callback = DownloadAttempt::on_success_callback;
+        using on_failure_callback = DownloadAttempt::on_failure_callback;
+
+        MirrorAttempt() = default;
+        explicit MirrorAttempt(Mirror* mirror);
+
+        void prepare_request(const Request& initial_request);
+        auto prepare_attempt(
+            CURLHandle& handle,
+            CURLMultiHandle& downloader,
+            const Context& context,
+            on_success_callback success,
+            on_failure_callback error
+        ) -> completion_function;
+
+        bool can_start_transfer() const;
+        bool has_failed() const;
+        bool has_finished() const;
+
+        void set_transfer_started();
+        void set_state(bool success);
+        void set_state(const Error& res);
+        void update_last_content(const Content* content);
+
+    private:
+
+        enum class State
+        {
+            WAITING_SEQUENCE_START,
+            PREPARING_DOWNLOAD,
+            RUNNING_DOWNLOAD,
+            LAST_REQUEST_FINISHED,
+            LAST_REQUEST_FAILED,
+            SEQUENCE_FINISHED,
+            SEQUENCE_FAILED,
```

Comment:
Nit: are we doing full caps for enum members?

Reply:
My understanding is that it's just for names of states, not specific to enums.

Reply:
It makes it consistent with the full caps used for constant (either in mamba or in its dependencies), and that's a convention I'm used too. Bu honestly I don't know if we have a convention for that in mamba, I just sticked with what was here before the PR.